### PR TITLE
event: Apply schedule event with preempt mode

### PR DESCRIPTION
**Pull request by honggyukim**
This patch adds EVENT_ID_PERF_SCHED_OUT_PREEMPT and use it to detect
whether the schedule event is because of preemption.

The example output is as follows.
```
  $ uftrace replay
  # DURATION     TID     FUNCTION
     1.620 us [  3595] | __monstartup();
     1.300 us [  3595] | __cxa_atexit();
              [  3595] | main() {
   284.645 us [  3595] |   fork();
              [  3595] |   wait() {
              [  3595] |     /* linux:sched-out */
              [  3597] |   } /* fork */
              [  3597] |   a() {
              [  3597] |     b() {
    38.621 us [  3597] |       /* linux:schedule (pre-empted) */
    11.080 us [  3597] |       /* linux:schedule (pre-empted) */
              [  3597] |       c() {
     5.580 us [  3597] |         getpid();
    11.381 us [  3597] |       } /* c */
   738.659 ms [  3597] |     } /* b */
   738.659 ms [  3597] |   } /* a */
   738.659 ms [  3597] | } /* main */
   739.806 ms [  3595] |     /* linux:sched-in */
   739.843 ms [  3595] |   } /* wait */
              [  3595] |   a() {
              [  3595] |     b() {
    29.460 us [  3595] |       /* linux:schedule (pre-empted) */
    15.000 us [  3595] |       /* linux:schedule (pre-empted) */
    10.001 us [  3595] |       /* linux:schedule (pre-empted) */
              [  3595] |       c() {
     3.940 us [  3595] |         getpid();
     6.860 us [  3595] |       } /* c */
   690.358 ms [  3595] |     } /* b */
   690.358 ms [  3595] |   } /* a */
     1.430  s [  3595] | } /* main */
```
Signed-off-by: Honggyu Kim <honggyu.kp@gmail.com>

### CODE REVIEW

**Comment by honggyukim** [comment]
This change is to show whether the schedule event occurred because of preemption. I slightly modified the `s-fork.c` example as follows.
```diff
diff --git a/tests/s-fork.c b/tests/s-fork.c
index fa30d2a6..9f74f569 100644
--- a/tests/s-fork.c
+++ b/tests/s-fork.c
@@ -21,6 +21,7 @@ static int b(void)

 static int c(void)
 {
+       for (int i = 0; i < 100000000; i++);
        return getpid() % 100000;
 }
```
The the replay output looks like this.
```
$ gcc -pg -o t-fork tests/s-fork.c
$ uftrace record t-fork
$ uftrace replay
# DURATION     TID     FUNCTION
   0.917 us [  5123] | __monstartup();
   0.583 us [  5123] | __cxa_atexit();
            [  5123] | main() {
 468.577 us [  5123] |   fork();
            [  5123] |   wait() {
            [  5123] |     /* linux:sched-out */
            [  5125] |   } /* fork */
            [  5125] |   a() {
            [  5125] |     b() {
            [  5125] |       c() {
  22.959 us [  5125] |         /* linux:schedule (preempted) */
  11.084 us [  5125] |         /* linux:schedule (preempted) */
   4.541 us [  5125] |         /* linux:schedule (preempted) */
   8.958 us [  5125] |         /* linux:schedule (preempted) */
  34.625 us [  5125] |         /* linux:schedule (preempted) */
   3.083 us [  5125] |         getpid();
 252.286 ms [  5125] |       } /* c */
 252.286 ms [  5125] |     } /* b */
 252.287 ms [  5125] |   } /* a */
 252.287 ms [  5125] | } /* main */
 254.817 ms [  5123] |     /* linux:sched-in */
 254.840 ms [  5123] |   } /* wait */
            [  5123] |   a() {
            [  5123] |     b() {
            [  5123] |       c() {
  12.333 us [  5123] |         /* linux:schedule (preempted) */
  44.166 us [  5123] |         /* linux:schedule (preempted) */
   5.041 us [  5123] |         getpid();
 221.061 ms [  5123] |       } /* c */
 221.061 ms [  5123] |     } /* b */
 221.061 ms [  5123] |   } /* a */
 476.396 ms [  5123] | } /* main */
```
The graph output is
```
$ uftrace graph
# Function Call Graph for 't-fork' (session: 8df36f71e8bb6e45)
========== FUNCTION CALL GRAPH ==========
# TOTAL TIME   FUNCTION
  728.685 ms : (1) t-fork
    0.917 us :  +-(1) __monstartup
             :  |
    0.583 us :  +-(1) __cxa_atexit
             :  |
  728.683 ms :  +-(1) main
  468.577 us :     +-(1) fork
             :     |
  254.840 ms :     +-(1) wait
  254.817 ms :     | (1) linux:schedule
             :     |
  473.349 ms :     +-(2) a
  473.348 ms :       (2) b
  473.348 ms :       (2) c
  138.666 us :        +-(7) liunx:schedule (preempted)
             :        |
    8.124 us :        +-(2) getpid
```
It works fine for `tui`, `script`, and `dump` commands as well. However, `report` output doesn't have a way to check if the schedule event is because of preemption and the output looks as follows.
```
$ uftrace report
  Total time   Self time       Calls  Function
  ==========  ==========  ==========  ====================
  728.683 ms   26.417 us           2  main
  473.349 ms    0.458 us           2  a
  473.348 ms    0.418 us           2  b
  473.348 ms  473.201 ms           2  c
  254.956 ms  254.956 ms           8  linux:schedule
  254.840 ms   22.292 us           1  wait
  468.577 us  468.577 us           2  fork
    8.124 us    8.124 us           2  getpid
    0.917 us    0.917 us           1  __monstartup
    0.583 us    0.583 us           1  __cxa_atexit
```
So I need to fix `report` command as well.

**Comment by namhyung** [comment]
`build_function_tree()` in cmds/report.c only checks `EVENT_ID_PERF_SCHED_IN` and adds a node using `sched_sym`.

**Comment by honggyukim** [comment]
@namhyung I've managed the report issue as well. Please have a look again. Thanks!
```
$ uftrace report
  Total time   Self time       Calls  Function
  ==========  ==========  ==========  ====================
    2.169  s   49.301 us           2  main
    1.429  s    0.880 us           2  a
    1.429  s    1.428  s           2  b
  739.843 ms   37.260 us           1  wait
  739.806 ms  739.806 ms           1  linux:schedule
  441.488 us  441.488 us          12  linux:schedule (pre-empted)
  284.645 us  284.645 us           2  fork
   18.241 us    8.721 us           2  c
    9.520 us    9.520 us           2  getpid
    1.620 us    1.620 us           1  __monstartup
    1.300 us    1.300 us           1  __cxa_atexit
```

**Comment by honggyukim** [comment]
I've just updated to support tui mode as well.

File: utils/perf.c [hunk 1]
```diff
@@ -597,7 +603,8 @@ void process_perf_event(struct uftrace_data *handle)
 			last = list_last_entry(&task->event_list.read, typeof(*last), list);
 
 			/* time filter is meaningful only for schedule events */
-			while (last->rstack.addr != EVENT_ID_PERF_SCHED_OUT) {
+			while (last->rstack.addr != EVENT_ID_PERF_SCHED_OUT ||
```
Comment (namhyung) [inline utils/perf.c:606]: `||` should be `&&`.

**Review by namhyung — COMMENTED** [review]
Can you check 208 watch_cpu test?

File: utils/perf.c (hunk 1 above)
Comment (honggyukim) [reply utils/perf.c]: Thanks. just updated.

**Review by namhyung — APPROVED** [review]
LGTM

**Comment by honggyukim** [comment]
@namhyung Thanks for the review. One last question is which one do you prefer between `linux:schedule (preempted)` and `linux:schedule (pre-empted)`?

**Comment by namhyung** [comment]
Doesn't matter but I slightly prefer the dash.

**Comment by honggyukim** [comment]
Then I will leave it as is.

**Pull request merged**